File: src/modules/graphs/1_Representation/AdjacencyMatrices.ts
/**
 * Cities: A, B, C, D, E
Routes:
A -> B
A -> C
B -> C
C -> D
D -> E
 */

class AdjacencyMatrices {
    public vertices: number;
    public matrix: number[][];
    constructor(vertices: number) {
        this.vertices = vertices;
        this.matrix = new Array(vertices).fill(0).map(() => new Array(vertices).fill(0))
    }
    addEdge(source: number, destination: number) {
        this.matrix[source][destination] = 1;
        this.matrix[destination][source] = 1;
    }
    displayMatrix() {
        for (let i = 0; i < this.vertices; i++) {
            console.log(this.matrix[i].join(' '));
        }
    }
}

const graph = new AdjacencyMatrices(5);

graph.addEdge(0, 1);
graph.addEdge(0, 2);
graph.addEdge(1, 3);
graph.addEdge(2, 4);

graph.displayMatrix();

export { }

File: src/modules/graphs/1_Representation/AdjacencyMatrices.go
package main

import "fmt"

type AdjacencyMatrices struct {
	vertices int
	matrix   [][]int
}

func newAdjacencyMatrices(vertices int) *AdjacencyMatrices {
	matrix := make([][]int, vertices)

	for i := range matrix {
		matrix[i] = make([]int, vertices)
	}

	return &AdjacencyMatrices{vertices: vertices, matrix: matrix}
}

func (am *AdjacencyMatrices) AddEdge(source, destination int) {
	if source >= am.vertices || destination >= am.vertices {
		fmt.Println("Error: Index out of bounds")
		return
	}
	am.matrix[source][destination] = 1
	am.matrix[destination][source] = 1
}

func (am *AdjacencyMatrices) displayMatrix() {
	for i := 0; i < am.vertices; i++ {
		fmt.Println(am.matrix[i])
	}
}

func main() {
	graph := newAdjacencyMatrices(5)

	graph.AddEdge(0, 1)
	graph.AddEdge(0, 2)
	graph.AddEdge(1, 3)
	graph.AddEdge(2, 4)

	graph.displayMatrix()
}
